### PR TITLE
[rl] Pin renderers 0.1.11, render with TorchTitan's tokenizer, take the renderer's typed config directly

### PR DESCRIPTION
**Pull request by felipemello1**
TLDR: 
- pin renderers
- Use our tokenizer and remove transformers dependency 
- Simplify code by using config classes directly rather than trying to map `model_name` -> config_type
- Simplify code by removing the need of registering muse glimmer

## Summary

TitanRL uses [`renderers`](https://github.com/PrimeIntellect-ai/renderers) to turn chat messages into token ids and parse completions back. Until now we wrapped it in our own `RendererConfig`:

```python
# before: recipe
renderer=RendererConfig(name="qwen3", enable_thinking=False)

# before: RendererConfig.build()
tokenizer = AutoTokenizer.from_pretrained(tokenizer_path)                  # transformers
args = {f.name: v for f in fields(self) if f.name in config_type.model_fields}   # silent drop
return create_renderer(tokenizer, config_type(**args))
```

**Problems:**

1. **Silent drops** (#4365). `build()` forwarded knobs by name-matching against the renderer's config and dropped the rest. When `renderers` replaced `preserve_all_thinking` with `thinking_retention`, the knob went inert with no error; two in-tree recipes were already hitting the same thing (`enable_thinking` on gpt-oss and Muse Glimmer, which have no such option).
2. **A second tokenizer.** `build()` loaded `transformers.AutoTokenizer` even though TorchTitan already has the tokenizer, and the controller then reached into `renderer._tokenizer` for `pad_id`.
3. **Transformers dependency**

**Solutions:**

`renderers==0.1.11` accepts a bring-your-own tokenizer and drops the transformers dependency:

1. **Recipes hold the library's typed config directly**: We don't have a wrapper anymore. We don't try to redirect 'qwen3` to the right config, e.g. `renderer=RendererConfig(name="qwen3", enable_thinking=False)`. Instead, we do:

   ```python
   from renderers import Qwen3RendererConfig, GptOssRendererConfig
   renderer=Qwen3RendererConfig(enable_thinking=False)
   renderer=GptOssRendererConfig(reasoning_effort="low")
   ```

2. **We load our own tokenizer:**`build_renderer` pairs that config with TorchTitan's tokenizer, through a small adapter (`RendererTokenizer`) that exposes the interface `renderers` expects.

   ```python
   # controller / rollout worker / generate.py
   tokenizer = HuggingFaceTokenizer(tokenizer_path=hf_assets_path)
   config.renderer = build_renderer(tokenizer=tokenizer, config=config.renderer)

   def build_renderer(*, tokenizer: HuggingFaceTokenizer, config: BaseRendererConfig) -> Renderer:
       if config.name == "auto": raise ValueError(...)      # unmatched local paths fall back unsafely
       if config.name == "default": raise ValueError(...)   # needs HF apply_chat_template
       renderer_tokenizer = RendererTokenizer(tokenizer)
       if isinstance(config, TorchTitanRendererConfig):     # renderer implemented in TorchTitan
           return config.renderer_cls(renderer_tokenizer, config)
       return create_renderer(tokenizer=renderer_tokenizer, config=config)
   ```

3. **We skip registration of new renderers***: Muse Glimmer's renderer lives in TorchTitan, so the renderer's registry cannot find it. We skip the need for the renderers registry that maps config -> renderer class. Check `TorchTitanRendererConfig`.

## Validation

- Alphabet sort, Qwen3-0.6B, 2 GPUs: validation reward 0.181 -> 0.548.

### CODE REVIEW

File: torchtitan/experiments/rl/tests/integration_tests.py [hunk 1]
```diff
@@ -46,7 +46,6 @@ def build_rl_test_list() -> list[OverrideDefinitions]:
                     "--async-loop.num-samples-per-prompt 2",
                     "--trainer.training.max_context_length 1024",
                     "--trainer.training.num_tokens_per_microbatch_per_dp_rank 2048",
-                    "--renderer.enable-thinking False",
```
Comment (wwwjn) [inline torchtitan/experiments/rl/tests/integration_tests.py:49]: Can we still pass CLI to disable thinking? Maybe this config can not be set by CLI because the base render class doesn't have this field, so we can not pass it.

I explicitly added enable-thinking=False for these CI tests because CI machine are easy to OOM (with 24GB memory), Can monitor if the CI works fine after removing it?

**Comment by HosseinKaviani-H** [comment]
Thanks for raising this. The silent drop is what actually impacted the work in #4145. The Muse Glimmer recipe sets enable_thinking=True but that renderer has no such field so it was dropped and the recipe read one way and ran another. Typed configs kill that whole class of bug. One caveat from the same experience: they catch a wrong name, not wrong behavior, so our renderer bugs all passed unit tests and only showed up in a real rollout. 

Here is Qwen3, which takes the create_renderer path. Is the TorchTitanRendererConfig path covered anywhere? A short rollout on the Muse Glimmer recipe would cover it.

File: torchtitan/config/configurable.py [hunk 2]
```diff
@@ -49,6 +49,8 @@ def to_dict(self) -> dict:
             def _convert(val):
                 if hasattr(val, "to_dict"):
                     return val.to_dict()
+                elif hasattr(val, "model_dump"):  # pydantic, e.g. renderer configs
+                    return _convert(val.model_dump())
```
Comment (tianyu-l) [inline torchtitan/config/configurable.py:53]: This is supposed to work with only native configs; o/w it'd be bloated eventually.

File: torchtitan/experiments/rl/rollout/rollouter.py [hunk 3]
```diff
@@ -241,11 +244,12 @@ def __init__(self, config: Config) -> None:
     async def setup_async(
         self,
         *,
-        renderer_config: RendererConfig,
+        renderer_config: BaseRendererConfig,
         hf_assets_path: str,
     ) -> None:
         """Build runtime dependencies after the worker actor is spawned."""
-        self._renderer = renderer_config.build(tokenizer_path=hf_assets_path)
+        tokenizer = HuggingFaceTokenizer(tokenizer_path=hf_assets_path)
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/rollout/rollouter.py:251]: this doesn't sound right -- it should be from tokenizer config build.

File: torchtitan/experiments/rl/generate.py [hunk 4]
```diff
@@ -31,13 +31,15 @@
...
 from torchtitan.components.checkpointer import CheckpointManager
+from torchtitan.components.tokenizer import HuggingFaceTokenizer
 from torchtitan.distributed.utils import set_batch_invariance
 from torchtitan.experiments.rl.examples.alphabet_sort import config_registry
 from torchtitan.experiments.rl.models.vllm_registry import (
     register_to_vllm,
     TORCHTITAN_CONFIG_FORMAT,
     TORCHTITAN_WORKER_CLS,
 )
+from torchtitan.experiments.rl.renderer import build_renderer
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/generate.py:42]: should be config.build

File: torchtitan/experiments/rl/renderer.py [hunk 5]
```diff
@@ -6,109 +6,156 @@
...
+            "Pick the model's renderer, e.g. Qwen3RendererConfig(...)."
+        )
+
+    renderer_tokenizer = RendererTokenizer(tokenizer)
+    if isinstance(config, TorchTitanRendererConfig):
+        return config.renderer_cls(renderer_tokenizer, config)
+    return create_renderer(tokenizer=renderer_tokenizer, config=config)
+
+
+class RendererTokenizer:
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/renderer.py:92]: Maybe
```suggestion
class RendererTokenizerWrapper:
```
to not confuse with `HuggingFaceTokenier`'s functionality

File: torchtitan/experiments/rl/renderer.py [hunk 6]
```diff
@@ -6,109 +6,156 @@
...
+        MuseGlimmerRendererConfig.renderer_cls = MuseGlimmerRenderer
+
+        build_renderer(tokenizer=tokenizer, config=MuseGlimmerRendererConfig())
+        # -> MuseGlimmerRenderer(RendererTokenizer(tokenizer), config), no registry involved
+    """
+
+    renderer_cls: ClassVar[type[Renderer]]
+
+
+def build_renderer(
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/renderer.py:45]: Instead of this, we should have a Renderer class and Renderer.Config in torchtitan and build from there. HuggingFaceTokenizer is the example https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/tokenizer.py#L94

What's the benefit? Well the main one is it can be swapped to be other renderer impl.

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 7]
```diff
@@ -18,6 +18,8 @@
 
 import dataclasses
 
+from renderers import Qwen3RendererConfig
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py:21]: I think this is because Prime-rl "happens" to be using similar way of configuring, but the "proper" and robust way to use something in other library is to build our own wrapper class.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py (hunk 7 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/examples/search_r1/config_registry.py:21]: See https://github.com/pytorch/torchtitan/pull/4356/changes#r3931881516

File: torchtitan/experiments/rl/renderer.py [hunk 8]
```diff
@@ -6,109 +6,142 @@
...
-    tool_parser: str | None = None
-    reasoning_parser: str | None = None
-    enable_thinking: bool | None = None
-    preserve_all_thinking: bool | None = None
-    preserve_thinking_between_tool_calls: bool | None = None
-
-    def build(self, *, tokenizer_path: str) -> Renderer:
-        # TODO(renderers#70): use TorchTitan's tokenizer once `renderers` supports
-        # bring-your-own-tokenizer (PR adds a Tokenizer protocol; drops transformers).
-        from transformers import AutoTokenizer
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/renderer.py:76]: pleasing

File: torchtitan/experiments/rl/renderer.py [hunk 9]
```diff
@@ -6,109 +6,142 @@
...
-            # so a register() call there never runs in that process.
-            from torchtitan.experiments.rl.models.muse_glimmer import (
-                renderer as _muse_glimmer_renderer,
+    renderers_config: Annotated[BaseRendererConfig, tyro.conf.Suppress]
+    """The library's typed config for the model, e.g. `Qwen3RendererConfig(enable_thinking=False)`.
+    Renderers and their options:
+    https://github.com/PrimeIntellect-ai/renderers/blob/renderers-v0.1.11/docs/renderer-config.md"""
+
+    def to_dict(self) -> dict[str, Any]:
+        return {"renderers_config": self.renderers_config.model_dump(mode="json")}
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/renderer.py:58]: oh nice

File: torchtitan/experiments/rl/models/muse_glimmer/renderer.py [hunk 10]
```diff
@@ -343,17 +334,13 @@ class _Piece(NamedTuple):
 
 class MuseGlimmerRenderer:
     def __init__(self, tokenizer, config: MuseGlimmerRendererConfig | None = None):
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/models/muse_glimmer/renderer.py:357]: add type for `tokenizer`

**Review by tianyu-l — APPROVED** [review]
looks wonderful
